lib/Parser/ptree.cpp

```diff
@@ -0,0 +1,254 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "ParserPch.h"
+
+ParseNodeUni * ParseNode::AsParseNodeUni()
+{
+    Assert(((this->Grfnop() & fnopUni) && this->nop != knopParamPattern) || this->nop == knopThrow);
+    return reinterpret_cast<ParseNodeUni *>(this);
```

Comment:
These patterns won't provide any additional runtime protection over the old way in release builds, is that okay?

Reply:
It is ok for now.  More work will need to be done to make those check cheaper, and also switch to the pattern of TryFrom() same as what we do for TryFromVar. Just want to check this in since the change is pretty disruptive already.